packages/autocomplete-js/src/__tests__/renderer.test.ts

```diff
@@ -74,13 +84,75 @@ describe('renderer', () => {
           },
         ];
       },
-      render({ createElement, Fragment }, root) {
+      render({ createElement, Fragment, render }, root) {
         expect(createElement).toBe(mockCreateElement);
         expect(Fragment).toBe(CustomFragment);
+        expect(render).toBe(mockRender);
         expect(mockCreateElement).toHaveBeenCalled();
 
         render(createElement(Fragment, null, 'testSource'), root);
       },
+      renderNoResults({ createElement, Fragment, render }, root) {
+        expect(createElement).toBe(mockCreateElement);
+        expect(Fragment).toBe(CustomFragment);
+        expect(render).toBe(mockRender);
+        expect(mockCreateElement).toHaveBeenCalled();
+
+        render(createElement(Fragment, null, 'testSource'), root);
+      },
+      renderer: {
+        createElement: mockCreateElement,
+        Fragment: CustomFragment,
+        render: mockRender,
+      },
+    });
+  });
+
+  test('sets `render` to `undefined` when not specified in the renderer', () => {
```

Comment:
The `renderer.render` option needs to stay optional not to break backwards compatibility. When users specify `renderer` but don't pass the `render` option, I think it's better to keep the provided `render` as `undefined` rather than default to Preact's `render`.

If we default to Preact's `render` and users use it, they might not realize they're using mismatched `createElement`/`Fragment` and `render` (e.g., React with Preact, Vue with Preact). I've seen it cause issues (see https://github.com/algolia/autocomplete/issues/886) but it can easily go unnoticed if you're not deep testing.

If we don't set `render`, it will be undefined. If users try to use it as a function, they'll have a runtime error which they'll immediately see while developing.

We could also help them out in development with a warning when we detect they've passed `renderer.createElement` or `renderer.Fragment`, but not `renderer.render`. wdyt?

Reply:
Everything else is defaulted so it'd make sense to default `render` as well imo. Having a possible `undefined` value increases complexity here.

Reply:
@francoischalifour If we go this route I think we should add a warning when we detect that they passed `renderer.createElement` or `renderer.Fragment` but not `renderer.render`, and they didn't specify the `render` option (meaning there's a mismatch).

People who pass a renderer typically use a build system so they have a dev mode.

Reply:
Updated in e8ad662f.